analyzers/src/SonarAnalyzer.CFG/ShimLayer/TypeInfoExtensions.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Copyright (c) Tunnel Vision Laboratories, LLC. All Rights Reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+namespace StyleCop.Analyzers.Lightup
+{
+    using System;
+    using Microsoft.CodeAnalysis;
+    using static System.Linq.Expressions.Expression;
+
+    public static class TypeInfoExtensions
+    {
+        private static readonly Func<TypeInfo, NullabilityInfo> ConvertedNullabilityAccessor = CreateNullabilityAccessor(nameof(ConvertedNullability));
+        private static readonly Func<TypeInfo, NullabilityInfo> NullabilityAccessor = CreateNullabilityAccessor(nameof(Nullability));
+
+        public static NullabilityInfo ConvertedNullability(this TypeInfo typeInfo) =>
+            ConvertedNullabilityAccessor(typeInfo);
+
+        public static NullabilityInfo Nullability(this TypeInfo typeInfo) =>
+            NullabilityAccessor(typeInfo);
+
+        private static Func<TypeInfo, NullabilityInfo> CreateNullabilityAccessor(string propertyName)
+        {
+            var typeInfoType = typeof(TypeInfo);
+            var property = typeInfoType.GetProperty(propertyName);
+            if (property == null)
+            {
+                return static _ => default;
+            }
+
+            Type nullableAnnotationType = typeof(NullableAnnotation), nullableFlowStateType = typeof(NullableFlowState);
+            var typeInfoParameter = Parameter(typeof(TypeInfo), "typeInfo");
+            var intermediateResult = Variable(property.PropertyType); // local variable which holds the Roslyn NullabilityInfo
+
+            // intermediateResult = typeInfo.{propertyName};
+            // return new Lightup.NullabilityInfo((Lightup.NullableAnnotation)intermediateResult.Annotation, (Lightup.NullableFlowState)intermediateResult.FlowState);
+            var body = Block(variables: new[] { intermediateResult },
+                Assign(intermediateResult, Property(typeInfoParameter, propertyName)),
+                New(typeof(NullabilityInfo).GetConstructor(new[] { nullableAnnotationType, nullableFlowStateType }),
+                    Convert(Property(intermediateResult, "Annotation"), nullableAnnotationType), // enum to enum conversion
+                    Convert(Property(intermediateResult, "FlowState"), nullableFlowStateType)));
+            var expression = Lambda<Func<TypeInfo, NullabilityInfo>>(body, typeInfoParameter);
+            return expression.Compile();
+        }
```

Comment:
This method should not be needed. `LightupHelpers.CreateSyntaxPropertyAccessor` inside a static constructor should do the trick for us. Something like in other files like this:
https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.CFG/ShimLayer/IFieldSymbolExtensions.cs

Reply:
I tried this before, but this wasn't working for several reasons:
* Several type conversions are needed to extract the information because we deal with a new struct and two new enums
* We would need to resort to `dynamic` and/or boxing with conversions to object if we don't create our own specialized expression tree
* My expression tree is highly efficient and does not allocate while also being concise.

Reply:
We should still follow the typical way how ShimLayer handles these. The more ve diverge, the more it will bite.

Since we need to return a special type, we should create a wrapper for that type. That means having `NullableInfoWrapper` instead of `NullableInfo`. Similar to `AnalyzerConfigOptionsWrapper` or `AnalyzerConfigOptionsProviderWrapper`. 
Then there should be a helper that will give us the wrapper. And wrapper will deal with the inner fields.

Reply:
> The more ve diverge, the more it will bite.

The code base is not consistent anyway. Look at e.g. `CaptureId`. `CreateOperationPropertyAccessor` special cases this one also and makes almost the same as what we do:

https://github.com/SonarSource/sonar-dotnet/blob/edde87f4d9ab5abc83bbf946a0b83970d375aeee/analyzers/src/SonarAnalyzer.CFG/ShimLayer/LightupHelpers.cs#L166-L175

I think it is more dangerous to take dependency to `CreateSyntaxPropertyAccessor` which is meant for SyntaxProperties than to create our own specialized solution.

>  Similar to AnalyzerConfigOptionsWrapper or AnalyzerConfigOptionsProviderWrapper

Both of these are classes while NullabilityInfo is a struct. We can expose it with zero overhead or with one boxing and two additional compiled expressions.

I don't see the point in creating a wrapper for a type that only carries two enums.

> NullableInfoWrapper instead of NullableInfo

If we want to be consequent with distinguishing between the shim types and the Roslyn types, we should also name the enums differently.